packages/fetchai/connections/ledger/contract_dispatcher.py

```diff
@@ -237,19 +248,80 @@ def get_raw_message(
         :param dialogue: the contract API dialogue
         :return: None
         """
-        contract = self.contract_registry.make(message.contract_id)
-        method_to_call = getattr(contract, message.callable)
-        try:
-            rm = method_to_call(api, message.contract_address, **message.kwargs.body)
-            response = ContractApiMessage(
+
+        def build_response(rm: bytes) -> ContractApiMessage:
+            return ContractApiMessage(
                 performative=ContractApiMessage.Performative.RAW_MESSAGE,
                 message_id=message.message_id + 1,
                 target=message.message_id,
                 dialogue_reference=dialogue.dialogue_label.dialogue_reference,
                 raw_message=RawMessage(message.ledger_id, rm),
             )
-            response.counterparty = message.counterparty
-            dialogue.update(response)
-        except Exception as e:  # pylint: disable=broad-except  # pragma: nocover
-            response = self.get_error_message(e, api, message, dialogue)
-        return response
+
+        return self._handle_request(api, message, dialogue, build_response)
+
+
+def _get_data(
+    api: LedgerApi, message: ContractApiMessage, contract: Contract,
+) -> bytes:
+    # first, check if the custom handler for this type of request has been implemented.
+    try:
+        assert message.performative.value in [
+            "get_state",
+            "get_raw_message",
+            "get_raw_transaction",
+            "get_deploy_transaction",
+        ]
+        method: Callable[[LedgerApi, ContractApiMessage], bytes] = getattr(
+            contract, message.performative.value
+        )
+        data = method(api, message)
+        return data
+    except NotImplementedError:
+        pass
+
+    # then, check if there is the handler for the provided callable.
+    method_to_call = getattr(contract, message.callable)
+    data = validate_and_call_callable(api, message, method_to_call)
+    return data
+
+
+def validate_and_call_callable(
+    api: LedgerApi, message: ContractApiMessage, method_to_call: Callable
+):
+    """
+    Validate a Contract callable, given the performative.
+
+    In particular:
+    - if the performative is either 'get_state' or 'get_raw_transaction', the signature
+      must accept ledger api as first argument and contract address as second argument,
+      plus keyword arguments.
+    - if the performative is either 'get_deploy_transaction' or 'get_raw_message', the signature
+      must accept ledger api as first argument, plus keyword arguments.
```

Comment:
Note; you only check quantity of args; not that they are actually including api and address

Reply:
the names can change; a check on the name would be useless.

Reply:
we know what the first one/two positional arguments should be though...

Reply:
I don't understand your comment; could you elaborate?